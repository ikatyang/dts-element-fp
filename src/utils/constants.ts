export const selectable_default = false;
export const placeholder_default = false;

export const get_interface_name_default = (index: number) =>
  `CurriedFunction${index + 1}`;

export const get_interface_generic_name_default = (index: number) =>
  `T${index + 1}`;
export const get_interface_generic_return_name_default = () => 'R';

export const get_interface_parameter_name_default = (index: number) =>
  `v${index + 1}`;
export const get_interface_parameter_placeholder_name_default = (
  index: number,
) => `_${index + 1}`;

export const get_placeholder_type_default = () => 'PH';

export const get_function_parameter_placeholder_name_default = (name: string) =>
  `_${name}`;
export const get_function_selectable_selector_name_default = () => '$SEL';
export const get_function_selectable_kind_name_default = () => '$KIND';
export const get_function_type_name_default = (name: string, mask: string) =>
  `${name}_${mask}`;
