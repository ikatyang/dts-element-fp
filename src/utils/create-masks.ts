import * as R from 'ramda';
import { left_pad } from './left-pad';

/**
 * - `masks[0]` -> `"000"`
 * - `masks[1]` -> `"100"`
 * - `masks[2]` -> `"010"`
 * - ...
 */
export const create_masks = (length: number) =>
  R.repeat(0, 2 ** length).map((_, index) =>
    left_pad(index.toString(2), '0', length)
      .split('')
      .reverse()
      .join(''),
  );
