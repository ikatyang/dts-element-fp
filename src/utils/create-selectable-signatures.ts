import * as dts from 'dts-element';
import {
  get_function_selectable_kind_name_default,
  get_function_selectable_selector_name_default,
} from './constants';

export interface CreateSelectableSignaturesOptions {
  get_selectable_kind_name?: () => string;
  get_selectable_selector_name?: () => string;
}

export interface CreateSelectableSignaturesInternalOptions
  extends CreateSelectableSignaturesOptions {
  /**
   * - normal: `undefined`
   * - various: `string`
   */
  group?: string;
  is_placeholder: (value: dts.IType) => boolean;
}

export function create_selectable_signatures(
  signatures: dts.IObjectMember[],
  options: CreateSelectableSignaturesInternalOptions,
): dts.IObjectMember[] {
  // istanbul ignore next
  const {
    group,
    is_placeholder,
    get_selectable_kind_name = get_function_selectable_kind_name_default,
    get_selectable_selector_name = get_function_selectable_selector_name_default,
  } = options;

  return signatures.map(signature => {
    const function_type = (signature.owned as dts.IFunctionDeclaration).type!;

    // istanbul ignore next
    const { parameters = [] } = function_type;

    const selectable_type = parameters
      .map(parameter => (is_placeholder(parameter.type!) ? '0' : '1'))
      .join('');

    return dts.create_object_member({
      owned: dts.create_function_declaration({
        name: undefined,
        type: dts.create_function_type({
          parameters: [],
          generics: [
            dts.create_generic_declaration({
              name: get_selectable_selector_name(),
              extends: dts.create_literal_type({ value: selectable_type }),
            }),
            ...(group === undefined
              ? []
              : [
                  dts.create_generic_declaration({
                    name: get_selectable_kind_name(),
                    extends: dts.create_literal_type({ value: group }),
                  }),
                ]),
          ],
          return: function_type,
        }),
      }),
    });
  });
}
