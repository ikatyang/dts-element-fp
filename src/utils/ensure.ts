export function ensure<T>(value: any, is: (x: any) => x is T): T {
  const is_something = is(value);
  // istanbul ignore next
  if (!is_something) {
    throw new Error('Invalid value');
  }
  return value;
}
