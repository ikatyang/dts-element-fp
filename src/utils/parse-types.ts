import * as dts from 'dts-element';

export function parse_types(
  types_string: string,
): Record<string, dts.IFunctionType> {
  const function_declarations = dts.parse(types_string)
    .members as dts.IFunctionDeclaration[];
  return function_declarations.reduce<Record<string, dts.IFunctionType>>(
    (current, function_declaration) => ({
      ...current,
      [function_declaration.name!]: function_declaration.type!,
    }),
    {},
  );
}
