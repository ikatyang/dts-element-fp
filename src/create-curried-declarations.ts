import * as dts from 'dts-element';
import {
  create_curried_types,
  CreateCurriedDeclarationsOptions,
} from './create-curried-types';
import { create_various_curried_types } from './create-curried-various-types';
import { get_placeholder_type_default } from './utils/constants';
import { sort_signatures } from './utils/sort-signatures';

export function create_curried_declarations(
  name: string,
  type: dts.IFunctionType | Record<string, dts.IFunctionType>,
  // istanbul ignore next
  options: CreateCurriedDeclarationsOptions = {},
): (dts.IVariableDeclaration | dts.ITypeDeclaration)[] {
  const type_declarations = dts.is_function_type(type)
    ? create_curried_types(name, type, options)
    : ((keys: string[]) =>
        keys.length === 1
          ? create_curried_types(name, type[keys[0]], options)
          : create_various_curried_types(name, type, options))(
        Object.keys(type),
      );

  // istanbul ignore next
  const { get_placeholder_type = get_placeholder_type_default } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const sorted_type_declarations = type_declarations.map(type_declaration => {
    const type_declaration_type = type_declaration.type;

    if (!dts.is_object_type(type_declaration_type)) {
      return type_declaration;
    }

    return {
      ...type_declaration,
      type: {
        ...type_declaration_type,
        members: sort_signatures(
          type_declaration_type.members as dts.IObjectMember[],
          { is_placeholder },
        ),
      },
    };
  });

  return [
    dts.create_variable_declaration({
      name,
      const: true,
      type: dts.create_general_type({
        name: sorted_type_declarations[0].name,
      }),
    }),
    ...sorted_type_declarations,
  ];
}
