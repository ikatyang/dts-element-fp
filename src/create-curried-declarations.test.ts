import * as dts from 'dts-element';
import { create_curried_declarations } from './create-curried-declarations';
import { CreateCurriedDeclarationsOptions } from './create-curried-types';
import { parse_types } from './utils/parse-types';

const test_cases = {
  normal: `
    function adjust<T, U>(fn: Morphism<T, U>, index: number, list: List<T>): (T | U)[];
  `,
  various: `
    function list<T, U>(fn: Morphism<T, U>, list: List<T>): U[];
    function functor<T, U>(fn: Morphism<T, U>, functor: Functor<T>): Functor<U>;
  `,
  various_mixed: `
    function list<T, U>(fn: Morphism<T, U>, list: List<T>): U[];
    function functor<T, U>(fn: Morphism<T, U>, functor: Functor<T>): Functor<U>;
    function mixed<T, U>(fn: Morphism<T, U>, target: List<T> | Functor<T>): U[] | Functor<U>;
  `,
  all_various: `
    function string(str: string): string;
    function list<T>(list: List<T>): T[];
    function mixed<T>(list: string | List<T>): string | T[];
  `,
  type_predicate: `
    function propIs<T, U extends {}, K extends string>(
      constructor: Constructor<T>,
      key: K,
      object: U,
    ): object is U & Record<K, T>;
  `,
};

it('should transform correctly with type = FunctionType', () => {
  expect(
    emit_curried_declarations(test_cases.normal, { use_first_type: true }),
  ).toMatchSnapshot();
});

it('should transform correctly with type = Record<string, FunctionType> (= 1 key)', () => {
  expect(emit_curried_declarations(test_cases.normal)).toMatchSnapshot();
});

it('should transform correctly with type = Record<string, FunctionType> (> 1 key)', () => {
  expect(emit_curried_declarations(test_cases.various)).toMatchSnapshot();
});

it('should omit unnecessary types with placeholder = false and type = FunctionType', () => {
  expect(
    emit_curried_declarations(test_cases.normal, { placeholder: false }),
  ).toMatchSnapshot();
});

it('should omit unnecessary types with placeholder = false and type = Record<string, FunctionType>', () => {
  expect(
    emit_curried_declarations(test_cases.various, { placeholder: false }),
  ).toMatchSnapshot();
});

it('should inline return type correctly with inline_return_type = true and type = FunctionType', () => {
  expect(
    emit_curried_declarations(test_cases.normal, { inline_return_type: true }),
  ).toMatchSnapshot();
});

it('should inline return type correctly with inline_return_type = true and type = Record<string, FunctionType>', () => {
  expect(
    emit_curried_declarations(test_cases.various, { inline_return_type: true }),
  ).toMatchSnapshot();
});

it('should inline return type correctly with inline_return_type = true and type = FunctionType contains type predicate', () => {
  expect(
    emit_curried_declarations(test_cases.type_predicate, {
      inline_return_type: true,
    }),
  ).toMatchSnapshot();
});

it('should inline return type correctly with inline_return_type = true and type = Record<string, FunctionType> contains all_various', () => {
  expect(
    emit_curried_declarations(test_cases.all_various, {
      inline_return_type: true,
    }),
  ).toMatchSnapshot();
});

it('should remove non-hoist kind in non-conflict types and remove hoist kind in selectable types with hoist_name = string  and type = Record<string, FunctionType> contains various_mixed', () => {
  expect(
    emit_curried_declarations(test_cases.various_mixed, {
      hoist_name: 'mixed',
    }),
  ).toMatchSnapshot();
});

it('should remove non-hoist kind in non-conflict types and remove hoist kind in selectable types with hoist_name = string  and type = Record<string, FunctionType> contains all_various', () => {
  expect(
    emit_curried_declarations(test_cases.all_various, {
      hoist_name: 'mixed',
    }),
  ).toMatchSnapshot();
});

function emit_curried_declarations(
  test_case: string,
  raw_options: CreateCurriedDeclarationsOptions & {
    use_first_type?: boolean;
  } = {},
) {
  const types = parse_types(test_case);

  const { use_first_type = false } = raw_options;

  const options = {
    selectable: true,
    placeholder: true,
    ...raw_options,
  };

  const curried_types = use_first_type
    ? create_curried_declarations('$', types[Object.keys(types)[0]], options)
    : create_curried_declarations('$', types, options);

  const top_level_element = dts.create_top_level_element({
    members: curried_types,
  });
  return dts.emit(top_level_element);
}
