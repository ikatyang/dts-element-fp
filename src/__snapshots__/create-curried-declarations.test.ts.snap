// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should inline return type correctly with inline_return_type = true and type = FunctionType 1`] = `
"declare const $: $_000;
type $_000 = {
    <T, U>(fn: Morphism<T, U>): $_100<T, U>;
    (_fn: PH, index: number): $_010;
    <T, U>(fn: Morphism<T, U>, index: number): $_110<T, U>;
    <T>(_fn: PH, _index: PH, list: List<T>): $_001<T>;
    <T>(_fn: PH, index: number, list: List<T>): $_011<T>;
    <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => (T | U)[];
    <$SEL extends \\"101\\">(): <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"011\\">(): <T>(_fn: PH, index: number, list: List<T>) => $_011<T>;
    <$SEL extends \\"001\\">(): <T>(_fn: PH, _index: PH, list: List<T>) => $_001<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => $_110<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_010;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_100<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): (T | U)[];
};
type $_100<T, U> = {
    (index: number): $_110<T, U>;
    (_index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => (T | U)[];
    <$SEL extends \\"01\\">(): (_index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => $_110<T, U>;
    (index: number, list: List<T>): (T | U)[];
};
type $_010 = {
    <T, U>(fn: Morphism<T, U>): $_110<T, U>;
    <T>(_fn: PH, list: List<T>): $_011<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => (T | U)[];
    <$SEL extends \\"01\\">(): <T>(_fn: PH, list: List<T>) => $_011<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_110<T, U>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): (T | U)[];
};
type $_110<T, U> = {
    (list: List<T>): (T | U)[];
};
type $_001<T> = {
    <U>(fn: Morphism<T, U>): $_101<T, U>;
    (_fn: PH, index: number): $_011<T>;
    <$SEL extends \\"11\\">(): <U>(fn: Morphism<T, U>, index: number) => (T | U)[];
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_011<T>;
    <$SEL extends \\"1\\">(): <U>(fn: Morphism<T, U>) => $_101<T, U>;
    <U>(fn: Morphism<T, U>, index: number): (T | U)[];
};
type $_101<T, U> = {
    (index: number): (T | U)[];
};
type $_011<T> = {
    <U>(fn: Morphism<T, U>): (T | U)[];
};
"
`;

exports[`should inline return type correctly with inline_return_type = true and type = FunctionType contains type predicate 1`] = `
"declare const $: $_000;
type $_000 = {
    <T>(constructor: Constructor<T>): $_100<T>;
    <K extends string>(_constructor: PH, key: K): $_010<K>;
    <T, K extends string>(constructor: Constructor<T>, key: K): $_110<T, K>;
    <U extends {}>(_constructor: PH, _key: PH, object: U): $_001<U>;
    <U extends {}, K extends string>(_constructor: PH, key: K, object: U): $_011<U, K>;
    <T, U extends {}>(constructor: Constructor<T>, _key: PH, object: U): $_101<T, U>;
    <$SEL extends \\"111\\">(): <T, U extends {}, K extends string>(constructor: Constructor<T>, key: K, object: U) => object is U & Record<K, T>;
    <$SEL extends \\"101\\">(): <T, U extends {}>(constructor: Constructor<T>, _key: PH, object: U) => $_101<T, U>;
    <$SEL extends \\"011\\">(): <U extends {}, K extends string>(_constructor: PH, key: K, object: U) => $_011<U, K>;
    <$SEL extends \\"001\\">(): <U extends {}>(_constructor: PH, _key: PH, object: U) => $_001<U>;
    <$SEL extends \\"11\\">(): <T, K extends string>(constructor: Constructor<T>, key: K) => $_110<T, K>;
    <$SEL extends \\"01\\">(): <K extends string>(_constructor: PH, key: K) => $_010<K>;
    <$SEL extends \\"1\\">(): <T>(constructor: Constructor<T>) => $_100<T>;
    <T, U extends {}, K extends string>(constructor: Constructor<T>, key: K, object: U): object is U & Record<K, T>;
};
type $_100<T> = {
    <K extends string>(key: K): $_110<T, K>;
    <U extends {}>(_key: PH, object: U): $_101<T, U>;
    <$SEL extends \\"11\\">(): <U extends {}, K extends string>(key: K, object: U) => object is U & Record<K, T>;
    <$SEL extends \\"01\\">(): <U extends {}>(_key: PH, object: U) => $_101<T, U>;
    <$SEL extends \\"1\\">(): <K extends string>(key: K) => $_110<T, K>;
    <U extends {}, K extends string>(key: K, object: U): object is U & Record<K, T>;
};
type $_010<K extends string> = {
    <T>(constructor: Constructor<T>): $_110<T, K>;
    <U extends {}>(_constructor: PH, object: U): $_011<U, K>;
    <$SEL extends \\"11\\">(): <T, U extends {}>(constructor: Constructor<T>, object: U) => object is U & Record<K, T>;
    <$SEL extends \\"01\\">(): <U extends {}>(_constructor: PH, object: U) => $_011<U, K>;
    <$SEL extends \\"1\\">(): <T>(constructor: Constructor<T>) => $_110<T, K>;
    <T, U extends {}>(constructor: Constructor<T>, object: U): object is U & Record<K, T>;
};
type $_110<T, K extends string> = {
    <U extends {}>(object: U): object is U & Record<K, T>;
};
type $_001<U extends {}> = {
    <T>(constructor: Constructor<T>): $_101<T, U>;
    <K extends string>(_constructor: PH, key: K): $_011<U, K>;
    <$SEL extends \\"11\\">(): <T, K extends string>(constructor: Constructor<T>, key: K) => boolean;
    <$SEL extends \\"01\\">(): <K extends string>(_constructor: PH, key: K) => $_011<U, K>;
    <$SEL extends \\"1\\">(): <T>(constructor: Constructor<T>) => $_101<T, U>;
    <T, K extends string>(constructor: Constructor<T>, key: K): boolean;
};
type $_101<T, U extends {}> = {
    <K extends string>(key: K): boolean;
};
type $_011<U extends {}, K extends string> = {
    <T>(constructor: Constructor<T>): boolean;
};
"
`;

exports[`should inline return type correctly with inline_return_type = true and type = Record<string, FunctionType> 1`] = `
"declare const $: $_00;
type $_00 = {
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): U[];
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => Functor<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => U[];
    <$SEL extends \\"01\\", $KIND extends \\"functor\\">(): <T>(_fn: PH, functor: Functor<T>) => $_functor_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"list\\">(): <T>(_fn: PH, list: List<T>) => $_list_01<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): Functor<U>;
};
type $_10<T, U> = {
    (list: List<T>): U[];
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => Functor<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => U[];
    (functor: Functor<T>): Functor<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): U[];
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): Functor<U>;
};
"
`;

exports[`should inline return type correctly with inline_return_type = true and type = Record<string, FunctionType> contains all_various 1`] = `
"declare const $: $_0;
type $_0 = {
    (str: string): string;
    <T>(list: List<T>): T[];
    <$SEL extends \\"1\\", $KIND extends \\"mixed\\">(): <T>(list: string | List<T>) => string | T[];
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => T[];
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => string;
    <T>(list: string | List<T>): string | T[];
};
"
`;

exports[`should omit unnecessary types with placeholder = false and type = FunctionType 1`] = `
"declare const $: $_000;
type $_000 = {
    <T, U>(fn: Morphism<T, U>): $_100<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number): $_110<T, U>;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => $_110<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_100<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): $_111<T, U>;
};
type $_100<T, U> = {
    (index: number): $_110<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => $_110<T, U>;
    (index: number, list: List<T>): $_111<T, U>;
};
type $_110<T, U> = {
    (list: List<T>): $_111<T, U>;
};
type $_111<T, U> = (T | U)[];
"
`;

exports[`should omit unnecessary types with placeholder = false and type = Record<string, FunctionType> 1`] = `
"declare const $: $_00;
type $_00 = {
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`should remove non-hoist kind in non-conflict types and remove hoist kind in selectable types with hoist_name = string  and type = Record<string, FunctionType> contains all_various 1`] = `
"declare const $: $_0;
type $_0 = {
    <$SEL extends \\"1\\", $KIND extends \\"mixed\\">(): <T>(list: string | List<T>) => $_mixed_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => $_list_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => $_string_1;
    <T>(list: string | List<T>): $_mixed_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T[];
type $_mixed_1<T> = string | T[];
"
`;

exports[`should remove non-hoist kind in non-conflict types and remove hoist kind in selectable types with hoist_name = string  and type = Record<string, FunctionType> contains various_mixed 1`] = `
"declare const $: $_00;
type $_00 = {
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <T>(_fn: PH, target: List<T> | Functor<T>): $_mixed_01<T>;
    <$SEL extends \\"11\\", $KIND extends \\"mixed\\">(): <T, U>(fn: Morphism<T, U>, target: List<T> | Functor<T>) => $_mixed_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <$SEL extends \\"01\\", $KIND extends \\"mixed\\">(): <T>(_fn: PH, target: List<T> | Functor<T>) => $_mixed_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"functor\\">(): <T>(_fn: PH, functor: Functor<T>) => $_functor_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"list\\">(): <T>(_fn: PH, list: List<T>) => $_list_01<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
    <T, U>(fn: Morphism<T, U>, target: List<T> | Functor<T>): $_mixed_11<U>;
};
type $_10<T, U> = {
    <$SEL extends \\"1\\", $KIND extends \\"mixed\\">(): (target: List<T> | Functor<T>) => $_mixed_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (target: List<T> | Functor<T>): $_mixed_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_mixed_01<T> = {
    <U>(fn: Morphism<T, U>): $_mixed_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
type $_mixed_11<U> = U[] | Functor<U>;
"
`;

exports[`should transform correctly with type = FunctionType 1`] = `
"declare const $: $_000;
type $_000 = {
    <T, U>(fn: Morphism<T, U>): $_100<T, U>;
    (_fn: PH, index: number): $_010;
    <T, U>(fn: Morphism<T, U>, index: number): $_110<T, U>;
    <T>(_fn: PH, _index: PH, list: List<T>): $_001<T>;
    <T>(_fn: PH, index: number, list: List<T>): $_011<T>;
    <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"101\\">(): <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"011\\">(): <T>(_fn: PH, index: number, list: List<T>) => $_011<T>;
    <$SEL extends \\"001\\">(): <T>(_fn: PH, _index: PH, list: List<T>) => $_001<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => $_110<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_010;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_100<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): $_111<T, U>;
};
type $_100<T, U> = {
    (index: number): $_110<T, U>;
    (_index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"01\\">(): (_index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => $_110<T, U>;
    (index: number, list: List<T>): $_111<T, U>;
};
type $_010 = {
    <T, U>(fn: Morphism<T, U>): $_110<T, U>;
    <T>(_fn: PH, list: List<T>): $_011<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"01\\">(): <T>(_fn: PH, list: List<T>) => $_011<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_110<T, U>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_111<T, U>;
};
type $_110<T, U> = {
    (list: List<T>): $_111<T, U>;
};
type $_001<T> = {
    <U>(fn: Morphism<T, U>): $_101<T, U>;
    (_fn: PH, index: number): $_011<T>;
    <$SEL extends \\"11\\">(): <U>(fn: Morphism<T, U>, index: number) => $_111<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_011<T>;
    <$SEL extends \\"1\\">(): <U>(fn: Morphism<T, U>) => $_101<T, U>;
    <U>(fn: Morphism<T, U>, index: number): $_111<T, U>;
};
type $_101<T, U> = {
    (index: number): $_111<T, U>;
};
type $_011<T> = {
    <U>(fn: Morphism<T, U>): $_111<T, U>;
};
type $_111<T, U> = (T | U)[];
"
`;

exports[`should transform correctly with type = Record<string, FunctionType> (= 1 key) 1`] = `
"declare const $: $_000;
type $_000 = {
    <T, U>(fn: Morphism<T, U>): $_100<T, U>;
    (_fn: PH, index: number): $_010;
    <T, U>(fn: Morphism<T, U>, index: number): $_110<T, U>;
    <T>(_fn: PH, _index: PH, list: List<T>): $_001<T>;
    <T>(_fn: PH, index: number, list: List<T>): $_011<T>;
    <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"101\\">(): <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"011\\">(): <T>(_fn: PH, index: number, list: List<T>) => $_011<T>;
    <$SEL extends \\"001\\">(): <T>(_fn: PH, _index: PH, list: List<T>) => $_001<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => $_110<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_010;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_100<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): $_111<T, U>;
};
type $_100<T, U> = {
    (index: number): $_110<T, U>;
    (_index: PH, list: List<T>): $_101<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"01\\">(): (_index: PH, list: List<T>) => $_101<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => $_110<T, U>;
    (index: number, list: List<T>): $_111<T, U>;
};
type $_010 = {
    <T, U>(fn: Morphism<T, U>): $_110<T, U>;
    <T>(_fn: PH, list: List<T>): $_011<T>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_111<T, U>;
    <$SEL extends \\"01\\">(): <T>(_fn: PH, list: List<T>) => $_011<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_110<T, U>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_111<T, U>;
};
type $_110<T, U> = {
    (list: List<T>): $_111<T, U>;
};
type $_001<T> = {
    <U>(fn: Morphism<T, U>): $_101<T, U>;
    (_fn: PH, index: number): $_011<T>;
    <$SEL extends \\"11\\">(): <U>(fn: Morphism<T, U>, index: number) => $_111<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => $_011<T>;
    <$SEL extends \\"1\\">(): <U>(fn: Morphism<T, U>) => $_101<T, U>;
    <U>(fn: Morphism<T, U>, index: number): $_111<T, U>;
};
type $_101<T, U> = {
    (index: number): $_111<T, U>;
};
type $_011<T> = {
    <U>(fn: Morphism<T, U>): $_111<T, U>;
};
type $_111<T, U> = (T | U)[];
"
`;

exports[`should transform correctly with type = Record<string, FunctionType> (> 1 key) 1`] = `
"declare const $: $_00;
type $_00 = {
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <$SEL extends \\"01\\", $KIND extends \\"functor\\">(): <T>(_fn: PH, functor: Functor<T>) => $_functor_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"list\\">(): <T>(_fn: PH, list: List<T>) => $_list_01<T>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;
